futurex_openedx_extensions/dashboard/serializers.py

```diff
@@ -362,3 +364,159 @@ def get_grade(self, obj: CourseOverview) -> Any:  # pylint: disable=no-self-use
             'letter_grade': course_grade.letter_grade,
             'is_passing': course_grade.passed,
         }
+
+
+class UserRolesSerializer(LearnerBasicDetailsSerializer):
+    """Serializer for user roles."""
+    tenants = serializers.SerializerMethodField()
+
+    def __new__(cls, *args: Any, **kwargs: Any) -> UserRolesSerializer:
+        """Create a new instance of the serializer."""
+        return super().__new__(cls, *args, **kwargs)
+
+    def __init__(self, instance: Any | None = None, data: Any = empty, **kwargs: Any):
+        """Initialize the serializer."""
+        self._org_tenant: dict[str, list[int]] = {}
+        self._roles_data: dict[Any, Any] = {}
+
+        permission_info = kwargs['context']['request'].fx_permission_info
+        self.orgs_filter = permission_info['view_allowed_full_access_orgs']
+        self.orgs_filter.extend(permission_info['view_allowed_course_access_orgs'])
+        self.permitted_tenant_ids = permission_info['permitted_tenant_ids']
+        self.query_params = self.parse_query_params(kwargs['context']['request'].query_params)
+
+        if instance:
+            self.construct_roles_data(instance if isinstance(instance, list) else [instance])
```

Comment:
what do think about this @OmarIthawi ? fetching one **page** data at a time in the serializer's initialization. Unlike `SerializerMethodField` which will fetch the data row by row. I'm thinking about applying this to other slow APIs if the `UserSignupSource` solution didn't improve them

Reply:
To understand correctly, you're optimizing a N+1 query.

So far we're not getting complaints about dashboard being slow, which means it's good enough. However, it doesn't mean we shouldn't improve, it just means it's not that urgent to do so.

I've checked the code quickly and I couldn't quickly grok it. We can discuss it in a call.